### DISCUSSION TITLE
[Cloud Sync] Receive API design to get data from Cloud

### DISCUSSION THREAD

**Discussion by suresh-lc** [lf-edge/edge-home-orchestration-go#492] [Q&A]
The MQTT client within Home Edge would receive the data whenever the data gets published to the MQTT broker at cloud. The communication between the service application and the Home Edge is Rest API based, like how we have designed for sending  data to Cloud. Since this is asynchronous process the time we receive the data at Home Edge is dynamic. There can be different design methods for sending the data to the service application. 

So main sub tasks would be :

1. Subscription API: Service application will call this Home Edge API to subscribe for a topic. We should initialize DataStorage at this point of time for topic of interest (e.g topic : home/living/lamp)

2. Broker would send event when there is publish for the topic. Home Edge MQTT client would receive the data and store in DataStorage. (e.g : topic : home/living/lamp, Message: Off)

3. Get Data API : Service Application would call the get data API of Home Edge to receive data for the topic. Internally data from DataStorage would be returned back.

 

>> Same Service application can register for multiple topics also. 

>> We update the old data values to save the recent data alone for any particular topic. 

 

If the above flow seems aggregable, we can start with developing the subscription API first. During the development we can see whether DataStorage is needed or a local Db would be enough. The reason for thought of local Db is that as we need not retain the history of data received for a particular topic and only maintain the latest data for the topic.

 

Came across two link which were good for MQTT, thought of sharing it here

1. http://www.steves-internet-guide.com/mqtt-publish-subscribe/

2. https://randomnerdtutorials.com/what-is-mqtt-and-how-it-works/

[CloudSync MQTT_Receive.pptx](https://github.com/lf-edge/edge-home-orchestration-go/files/8108314/CloudSync.MQTT_Receive.pptx)

**Comment by suresh-lc** [comment]
Base version:
1. Service Application calls subscribe API and Home Edge subscribes to the topic
2. Home Edge receives the publish topics and save to local Db or variable
3. Latest value for the topics would be saved in Home Edge
4. Service Application can call the Get API to get the latest value for the subscribed topic

**Reply by tiokim** [reply]
It's great idea. I think Service Application can get the latest value of topic through JSON format.
